### PR TITLE
fix: bail early if no printers on the network

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/22409.

Fixes an edge case in the event that no valid printers are known to a given device.

cc @nornagon @deepak1556 @zcbenz 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed a potential crash on devices which had not connected any printers to their network.

### CODE REVIEW

File: shell/browser/api/electron_api_web_contents.cc [hunk 1]
```diff
@@ -1763,8 +1763,17 @@ void WebContents::OnGetDefaultPrinter(
     return;
   }
 
-  base::string16 printer_name =
-      device_name.empty() ? default_printer : device_name;
+  base::string16 printer_name;
```
Comment (deepak1556) [inline shell/browser/api/electron_api_web_contents.cc]: Can you remove https://github.com/electron/electron/blob/master/shell/browser/api/electron_api_web_contents.cc#L1848-L1856 and consolidate this block too,

```
base::string16 printer_name =
      device_name.empty() ? default_printer : device_name;
// If there are no valid printers available on the network, bail.
if (printer_name.empty() || !IsDeviceNameValid(printer_name)) {
   std::move(print_callback).Run(false, "no valid printers available");
   return;
 }
```

The current condition seems to omit device_name always.

**Comment by thomasdao** [comment]
Hi @codebytere, thanks for working on this. Just wanna ask if printer is not available, will Electron still show the print dialog? I tried print with `Preview` on Mac and although I don't have any printer available, the print dialog still shows up for me. I hope Electron would show the dialog, but if it's difficult I think fixing the crash is definitely more important.

**Comment by codebytere** [comment]
@thomasdao it's difficult enough that for now i'm just going to mitigate the crash, but I can spend a bit of time in the next few weeks searching for a more robust solution!

File: shell/browser/api/electron_api_web_contents.cc [hunk 2]
```diff
@@ -1765,6 +1765,13 @@ void WebContents::OnGetDefaultPrinter(
 
   base::string16 printer_name =
       device_name.empty() ? default_printer : device_name;
+
+  // If there are no valid printers available on the network, bail.
+  if (printer_name.empty() || !IsDeviceNameValid(printer_name)) {
```
Comment (deepak1556) [inline shell/browser/api/electron_api_web_contents.cc]: https://github.com/electron/electron/blob/master/shell/browser/api/electron_api_web_contents.cc#L1851-L1856 check is now redundant, only advantage is bailing early which doesn't save much, we can remove that.

File: shell/browser/api/electron_api_web_contents.cc [hunk 3]
```diff
@@ -1765,6 +1765,15 @@ void WebContents::OnGetDefaultPrinter(
 
   base::string16 printer_name =
       device_name.empty() ? default_printer : device_name;
+
+  // We set the default to the system's default printer and only update
+  // if at the Chromium level if the user overrides.
+  // If there are no valid printers available on the network, we bail.
+  if (printer_name.empty() || !IsDeviceNameValid(printer_name)) {
+    std::move(print_callback).Run(false, "no valid printers available");
```
Comment (zcbenz) [inline shell/browser/api/electron_api_web_contents.cc]: `print_callback` might be `null` if user does not pass an callback, should check `if (print_callback)` here.

**Comment by zcbenz** [comment]
The tests are failing on macOS.

File: shell/browser/api/electron_api_web_contents.cc (hunk 2 above)
Comment (codebytere) [reply shell/browser/api/electron_api_web_contents.cc]: @deepak1556 hmm - i'm rethinking this in light of the newly failing test. I think there is in fact value in telling the user that _they_ passed in an invalid printer independent of printers being available on the network. One throws an error and the other calls a callback, and i think as a rule user-error should be surfaced early with a true Error and not just a callback.

Comment (zcbenz) [reply shell/browser/api/electron_api_web_contents.cc]: I'm good with either solution. But we should explicitly document that the `print` throws on invalid input, currently the API design implies that all errors are passed via callback.

Comment (deepak1556) [reply shell/browser/api/electron_api_web_contents.cc]: I am fine with it as well, was suggesting this because errors can be raised through callbacks.

Comment (codebytere) [reply shell/browser/api/electron_api_web_contents.cc]: done!
